cmd/evm/staterunner.go

```diff
@@ -100,18 +100,19 @@ func runStateTest(fname string, cfg vm.Config, jsonOut, dump bool) error {
 		for _, st := range test.Subtests() {
 			// Run the test and aggregate the result
 			result := &StatetestResult{Name: key, Fork: st.Fork, Pass: true}
-			test.Run(st, cfg, false, rawdb.HashScheme, func(err error, snaps *snapshot.Tree, state *state.StateDB) {
-				if state != nil {
-					root := state.IntermediateRoot(false)
+			test.Run(st, cfg, false, rawdb.HashScheme, func(err error, snaps *snapshot.Tree, statedb *state.StateDB) {
+				var root common.Hash
+				if statedb != nil {
+					root = statedb.IntermediateRoot(false)
 					result.Root = &root
 					if jsonOut {
 						fmt.Fprintf(os.Stderr, "{\"stateRoot\": \"%#x\"}\n", root)
 					}
-				}
-				// Dump any state to aid debugging
-				if dump {
-					dump := state.RawDump(nil)
-					result.State = &dump
+					if dump { // Dump any state to aid debugging
```

Comment:
On phone, but. You don't see it here, but external code calls commit prior to the callback. So the state we get here is already comitted. 

So that is why it works. The intermediate root is just to obtain the root, not for side effects